IoHandle.md

```diff
@@ -0,0 +1,387 @@
+   virtual IoCallResult applyOptions() PURE;
+
+   virtual void setTransport(IoTransportPtr transport) PURE; 
+   virtual IoTransportPtr transport() PURE;
+
+   // Calls made from Envoy::Network::Address::Instance
+   virtual IoCallResult bind() PURE;
+   virtual IoCallResult connect() PURE; 
+   virtual IoCallResult close() PURE;
+
+   // Read/Write operations
+   virtual IoCallResult read(Buffer::Instance& buffer) PURE;
+   virtual IoCallResult write(Buffer::Instance& buffer, bool end_stream) PURE;
+
+   // Descriptor
+   const IoDescriptor& descriptor() const PURE;
+};
+
+class IoSocketHandle : public IoHandle {
+public:
+   IoSocketHandle(const IoDescriptor& descriptor) : descriptor_(descriptor) { }
+   IoSocketHandle(const IoDescriptor& descriptor, const IoAddress& local_address)
+     :descriptor_(descriptor), 
+      local_addr_(local_addr){ }
+	
+   IoSocketHandle(const IoDescriptor& descriptor, const IoAddress& local_addr, 
+        const IoAddress& remote_addr)  : descriptor_(descriptor),
+        local_addr_(local_addr),  remote_addr_(remote_addr) { }
+   virtual IoSocketHandle()  { }
+   virtual ~IoSocketHandle() { }
+
+   // I/O Options API
+   virtual IoCallResult setOption(IoOptionName name, IoOptionValue value)  override ;
+   virtual const IoOptionValue& option(IoOptionName name)  override ;
+   virtual size_t  options(vector<std::pair<IoOptionName,IoOptionValue>>& options)  override ;
+   virtual IoCallResult applyOptions()  override ;
+   virtual void setTransport(IoTransportPtr transport) override {
+             transport_socket_ = transport);
+   virtual IoTransportPtr transport() override { return transport_socket_; }
+
+   // Operations in Envoy::Network::Address::Instance
+   virtual IoCallResult  bind()  override ;
+   virtual IoCallResult connect()  override ;
+   virtual IoCallResult close() override;
+   virtual IoCallResult socket() PURE;
+
+   // Read/Write operations
+   virtual IoCallResult read(Buffer::Instance& buffer) override ;
+   virtual IoCallResult write(Buffer::Instance& buffer, bool end_stream) override ;
+
+   const IoDescriptor& descriptor() const  override { return(descriptor_); }
+
+   // Addresses (see include/network/address.h below)
+   const IoAdress&	localAddress() const { return(local_addr_); }
+   const IoAdress&	remoteAddress() const { return(remote_addr_); }
+   bool	 hasRemoteAddr() const { return !remote_addr_.isEmpty(); }
+
+protected:
+   IoSocketDescriptor descriptor_;
+   IoAddress local_addr_;
+   IoAddress	 remote_addr_;
+   IoTransportSocket transport_socket_;
+};
+
+```
+
+Similarly, we would define a new class IoHandleVpp that would accommodate differences required to interface to VPP. And it would have a customized IoTransportVpp member.
+
+
+
+###### IoAddress and derivative classes
+
+For the address classes, the intention was to remove all the methods using the fd. The IoAddress class is intended to replace the Instance class defined in address.h.
```

Comment:
Per my comment on transport socket below, I think what we want to be doing is trying to reuse the existing classes as much as possible and only swapping out fd access with 'fd wrapper' access, i.e. removing the int with our class which abstracts away fd functionality.  

Moving all use of address over to something new will cause a ton of churn.  Simply replacing the functions we want to replace seems like a more tenable project to tackle, and meshes with the incremental approach I think we're going to have to take in any case :-)

Reply:
@alyssawilk: I hear you- thanks for the clarity. As I said, I will take this approach.

Reply:
+1

Reply:
And to add on, I think it should be easy to do this if we have Addresses spit out of the existing dispatcher/connection/etc. where we can hide the underlying handles easily. That's what I was suggesting on the call. If we do this, I don't think we need to change the existing classes very much or at all.

Reply:
So if I understand correctly, we want IoHander to encapsulate opaquely the underlying implementation (for example, a ```BSDSockeIoHandle{}``` would encapsulate the fd, but never expose it).

And then things like:

```
IoHandle->bind()
IoHandle->connect()
IoHandle->listen()
IoHandler->read()  // To be used by the TransportSocket, not to be used directly
IoHandle->write() // To be used by the TransportSocket, not to be used directly
IoHandle->setNoDelay() // presuming we go with @mattklein123 's suggestion for options
```
etc to be exposed.  So  a VPPIoHandler might do something completely different internal to itself to accomplish these ends than a BSDSocketIoHandle or a QUICIoHandle.

We can then go to all the places that had been calling fd(), and replace the direct calls that are using the return from fd() with calls to the analogous IoHandle method.

Reply:
So would it make sense then to do this in small step PRs like:

1. Write an IoHandle (and BSDSocketIOHandle) with the aforementioned methods
2. A PR to replace use of fd() with IOHandle

That way we can sort out the shape of an IoHandle and then start with the minutia of replacing all the fd() calls and things using fds directly :)

Reply:
Yep, that SGTM. And then we just set up the inheritance as mentioned above so if folks want to add their own IoHandlerImpl subclasses they can do so.

Reply:
> Er, your first comment SGTM. My suggestion was to start even smaller - implementing IoHandle fd() and update relevant classes that took an fd() to take an IoHandle, but if swapping bind/connect/listen/read/write isn't too large you can totally start with something a bit more useful :-)

Just to make sure I understand correctly... in the final steady state we don't want an IoHandle->fd() correct?  Would be prone to abuse :)

Reply:
Yes exactly. @alyssawilk is suggesting a way to do this incrementally so we can have small PRs. I agree with her proposal.

Reply:
Yep, you got it.  I'm out tomorrow but happy to touch base Monday on slack or vc if you want to sanity check anything else in real time.